sql: adding parallel test framework

RaduBerinde opened the pull request:
Adding a test orchestrator on top of logic test, capable of "firing off"
multiple test files in parallel. Using this framework, we add a test that
exposes issue #4472. @tschottdorf @petermattis 

The first commit only moves the logictest data to a subdir; it should be easier to review just the second commit. If you'd prefer to leave them as is and put the new tests in a different subdir, e.g. `sql/partestdata`, let me know.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4516)

<!-- Reviewable:end -->

andreimatei commented:
LGTM

Maybe consider adding a numeric field to the `run` command to specify how much parallelism there should be for each file (e.g. `run 10 reader 10 writer`), and a simple label and GOTO command so you can create loops (to remove the repetition in the test file). Maybe labels can act as counters too. E.g. 
```
label start

statement ok
INSERT INTO T VALUES ((SELECT MAX(k+1) FROM T))

goto start 100
```

---

Review status: 0 of 55 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516#-:-KAuCBv7k_Si9iDUq_mu:759863735)*
<!-- Sent from Reviewable.io -->

petermattis commented:
`testdata/logictest` and `testdata/paralleltest` exhibit a bit of stuttering. Rather than another level of hierarchy, I think I'd move the parallel test data to `pardata` or `partestdata` or something like that. This also matches up with the `tracedata` directory.

@andreimatei I'd like to avoid gotos, but see my comment about the ability to repeat a statement or query N times.

LGTM

---

Reviewed 99 of 99 files at r1.
Review status: 49 of 55 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/testdata/paralleltest/subquery_retry/txn, line 5 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516#-KAuv5P95kC8Bqs_wPbK:-KAuv5P95kC8Bqs_wPbL:1256755842) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5b0170b363a607b68928c53eda3dcadcaeececf/sql/testdata/paralleltest/subquery_retry/txn#L5)):*
Rather than repeating the statement so many times, perhaps there should be an extension to the normal logic test so that we can repeat a command N times. Something like:

```
repeat 100
statement ok
...
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516#-:-KAupcGxcLh-mYGtTTuK:1218191387)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Left the current files where they are, moved the new test to `partestdata`, and added `repeat` statement in logic_test.

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516#-:-KAv0cc5EGdCtfBXfEcZ:127294363)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/logic_test.go, line 319 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516#-KAv5DfTutqja8_w6m5m:-KAv5DfUUar6GKpYT3b_:7469332) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fbb03bb542471f57d9fe588bb5f3dc53c835ddb/sql/logic_test.go#L319)):*
I would think you should reset `repeat = 1` regardless of whether the statement/query was executed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/logic_test.go, line 319 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516#-KAv5DfTutqja8_w6m5m:-KAvBjKVm7YIIg_zVpy6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fbb03bb542471f57d9fe588bb5f3dc53c835ddb/sql/logic_test.go#L319)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516#-:-KAvDbr6KlP6332BIHep:676834233)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: adding parallel_test

Adding a test orchestrator on top of logic test, capable of "firing off"
multiple test files in parallel. Using this framework, we add a test that
exposes issue #4472.

Also adding a "repeat" instruction to logic test which causes the following
statement or query to be repeated the given number of times.

tschottdorf commented:
LGTM. More limited in focus than what I had in mind, but probably useful for some things. I think a slim go library to create clients that execute individual testdata snippets would be useful, especially looking at tests which can also run against "real" clusters. But this is a good start.

---

Reviewed 9 of 108 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516#-:-KAvP_IgS2sZIJt3oYYn:1853482981)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks. Yeah the intention is to extend it as needed, either by making the "script" more expressive, or allowing most of this to be leveraged programatically (or both).

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4516#-:-KAvVpj2myiEfMyaMgAk:-1680750370)*
<!-- Sent from Reviewable.io -->
